bazel/aar_with_jni.bzl

```
@@ -4,82 +4,214 @@ load("@google_bazel_common//tools/maven:pom_file.bzl", "pom_file")
+        out_javadocs=$$4
+
+        echo "Outputting pom.xml, sources.jar, and javadocs.jar..."
+        cp $$src_aar $$out_aar
+        cp $$src_pom_xml $$out_pom_xml
+        cp $$src_sources_jar $$out_sources_jar
+        cp $$src_javadocs $$out_javadocs
+        echo "Finished!"
+        """,
+    )
+
+def _create_aar(name, archive_name, classes_jar, jni_archive, proguard_rules, visibility):
+    """
+    This macro rule manually creates an aar artifact.
+
+    The underlying gen rule does the following:
+    1. Create the final aar manifest file.
+    2. Unzips the apk file generated by the `jni_archive_name` into a temporary directory.
+    3. Renames the `lib` directory to `jni` directory since the aar requires the so files
+       to be in the `jni` directory.
+    4. Copy the android binary `jar` output from the `android_binary_name` as `classes.jar`.
+    5. Copy the proguard rules specified in the macro parameters.
+    6. Override the apk's aar with a generated one.
+    7. Zip everything in the temporary directory into the output.
+
+
+    :param name Name of the aar generation rule.
+    :param archive_name Name of the resulting aar archive.
+    :param classes_jar The classes.jar file which contains all the kotlin/java classes.
+    :param jni_archive The apk with the desired jni libraries.
+    :param proguard_rules The proguard.txt file.
+    :param visibility The bazel visibility for the underlying rule.
+    """
+    _aar_output = name + "_local.aar"
+
+    # This is to generate the envoy mobile aar AndroidManifest.xml
+    _manifest_name = name + "_android_manifest"
+    native.genrule(
+        name = _manifest_name,
+        outs = [_manifest_name + ".xml"],
+        cmd = "cat > $(OUTS) <<EOF {}EOF".format(_manifest("io.envoyproxy.envoymobile")),
     )
 
-    # Create a dummy manifest file for our android_binary
     native.genrule(
-        name = archive_name + "_binary_manifest_generator",
-        outs = [archive_name + "_generated_AndroidManifest.xml"],
+        name = name + "_only_aar",
+        outs = [_aar_output],
+        srcs = [
+            classes_jar,
+            jni_archive,
+            _manifest_name,
+            proguard_rules,
+        ],
         cmd = """
-cat > $(OUTS) <<EOF
-<manifest
-  xmlns:android="http://schemas.android.com/apk/res/android"
-  package="does.not.matter">
-  <uses-sdk android:minSdkVersion="999"/>
-</manifest>
-EOF
-""",
+        # Set source variables
+        set -- $(SRCS)
+        src_classes_jar=$$1
+        src_jni_archive_apk=$$2
+        src_manifest_xml=$$3
+        src_proguard_txt=$$4
+
+        original_directory=$$PWD
+
+        echo "Constructing aar..."
+        final_dir=$$(mktemp -d)
+        cp $$src_classes_jar $$final_dir/classes.jar
+        cd $$final_dir
+        unzip $$original_directory/$$src_jni_archive_apk > /dev/null
+        if [[ -d lib ]]; then
+            mv lib jni
+        else
+            echo "No jni directory found"
+        fi
```

```suggestion
            echo "No jni directory found"
            exit 1
        fi
```

might want to fail here

---

It was to allow for the optionality of not having a native dependency (helped with testing this bazel rule)